### DISCUSSION TITLE
Args/ArgTypes in a Monorepo Environment

### DISCUSSION THREAD

**Discussion by stevegreco** [storybookjs/storybook#28556] [Help]
### Summary

I am curious if anyone has any experience using Storybook in a Monorepo environment.  I currently have a project using Turborepo and it is similar to the official [example from turborepo](https://github.com/vercel/turbo/tree/main/examples/design-system)

What I am noticing is that when Storybook is using the bundled and transpiled component code, it can no longer figure out the Arg Types, so we are finding ourselves having to manually set `argTypes` for every single arg.  The Story files are properly getting the types and the args/argTypes fields are able to determine our props perfectly, but it does not translate to the actual controls and docs blocks.

One example would be

```ts
interface Props {
  size: 'sm' | 'md' | 'lg'
}
```

This ends up just showing up in Storybook as a `string` type, so we are not able to see the available options, and the control maps to a basic text input.

Does anyone have any experience with this type of thing that could provide some feedback on how to set up an environment like this?

### Additional information

```ts
const path = require('path');

const config = {
  framework: {
    name: '@storybook/react-webpack5',
    options: {},
  },

  stories: ['../src/**/*.stories.(tsx|mdx)', '../src/**/*.mdx'],

  addons: [
    '@storybook/addon-links',
    {
      name: '@storybook/addon-essentials',
      options: {
        actions: true,
        backgrounds: true,
        controls: true,
        docs: true,
        toolbars: true,
        viewport: true,
      },
    },
    '@storybook/addon-a11y',
    '@storybook/addon-toolbars',
    '@storybook/addon-mdx-gfm',
    '@storybook/addon-webpack5-compiler-swc',
    '@storybook/addon-interactions',
  ],
  typescript: {
    check: true,
    reactDocgen: 'react-docgen-typescript',
    reactDocgenTypescriptOptions: {
      shouldExtractLiteralValuesFromEnum: true,
      shouldExtractValuesFromUnion: true,
      shouldIncludePropTagMap: true,
    },
    checkOptions: {
      typescript: {
        tsconfig: 'tsconfig.storybook.json',
      },
    },
  },
};
```

**Comment by Marklb** [comment]
Is there a reason that you need to use the transpiled files? ~I haven't run the example that you linked, to confirm this, but it looks like the example's Storybook is configured to use the source files.~

~The following, in their `.storybook/main.js` shows an alias to replace `"ui"` with `resolve(__dirname, "../../../packages/ui/")`. *I would have expected the alias to replace `"@acme/ui"` with `resolve(__dirname, "../../../packages/ui/src")`, but I don't have much experience with turborepo or vite, so I may just not know how those aliases work.*
https://github.com/vercel/turbo/blob/3eb27d6664a3efb4177d4e40dfd3340b0c7fa2dd/examples/design-system/apps/docs/.storybook/main.js#L20-L34~

I started thinking about this again and realized the alias snippet that I linked is actually doing what you described and pointing to the built files, I am pretty sure. So, my answer was probably wrong, but I still question why a documentation tool would be pointing to files that may have documentation details removed. Maybe it is fine, as long as the generated types files include everything, such as comments. I don't know enough about reactDocgen to answer that though, so you may need someone else to answer that.

**Comment by astjohnpaycor** [comment]
We are having this same issue with our monorepo, and the only solution we've been able to come up with so far is in the story files, we import the components from their original source instead of the compiled package.

**Reply by stevegreco** [reply]
That is what we are doing.  A lot of my issues seem to be coming from the use of `forwardRef`, this seems to be breaking some of the auto-genrated ArgType tables.  Whatever is currently being used for docgen, seems to really only work with very simple types.
